### PR TITLE
Reroute Symmetric/Hermitian + Diagonal through triangular

### PR DESCRIPTION
**Pull request by jishnub**
This should fix the `Diagonal`-related issue from https://github.com/JuliaLang/LinearAlgebra.jl/issues/1089, although the `SymTridiagonal` one still remains.
```julia
julia> using LinearAlgebra

julia> a = Matrix{BigFloat}(undef, 2,2)
2×2 Matrix{BigFloat}:
 #undef  #undef
 #undef  #undef

julia> a[1] = 1; a[3] = 1; a[4] = 1
1

julia> a = Hermitian(a)
2×2 Hermitian{BigFloat, Matrix{BigFloat}}:
 1.0  1.0
 1.0  1.0

julia> b = Symmetric(a)
2×2 Symmetric{BigFloat, Matrix{BigFloat}}:
 1.0  1.0
 1.0  1.0

julia> c = Diagonal([1,1])
2×2 Diagonal{Int64, Vector{Int64}}:
 1  ⋅
 ⋅  1

julia> a+c
2×2 Hermitian{BigFloat, Matrix{BigFloat}}:
 2.0  1.0
 1.0  2.0

julia> b+c
2×2 Symmetric{BigFloat, Matrix{BigFloat}}:
 2.0  1.0
 1.0  2.0
```

### CODE REVIEW

**Comment by araujoms** [comment]
Shouldn't it be possible to write a universal fallback? Writing a dedicate sum function for every wrapper seems Sisyphean.

**Comment by jishnub** [comment]
Yes, this should be possible by defining broadcast styles, but that's not there yet

File: stdlib/LinearAlgebra/src/special.jl [hunk 1]
```diff
@@ -288,6 +288,25 @@ function (-)(A::UniformScaling, B::Diagonal)
     Diagonal(Ref(A) .- B.diag)
 end
 
+for f in (:+, :-)
+    @eval function $f(D::Diagonal{<:Number}, S::Symmetric)
+        uplo = sym_uplo(S.uplo)
+        return Symmetric(parentof_applytri($f, Symmetric(D, uplo), S), uplo)
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/special.jl:294]: The superfluous `Symmetric` wrapper around `D` is peeled off by `parentof_applytri`, right?

Comment (jishnub) [reply stdlib/LinearAlgebra/src/special.jl:294]: Yes, the wrapper is added only to dispatch to the correct method. The actual computation happens with a `Diagonal`. The sequence is `Symmetric(D) -> applytri -> uppertriangular(D) -> D`

**Review by dkarrasch — APPROVED** [review]

**Comment by KristofferC** [comment]
Seems to fail when backported: https://buildkite.com/julialang/julia-release-1-dot-11/builds/196#0191d678-ee90-4570-bb87-45c26cfbf26d

```
 Expression: H + D == D + H == HM + D
  UndefRefError: access to undefined reference
  Stacktrace:
    [1] getindex
      @ ./essentials.jl:916 [inlined]
    [2] getindex
      @ ./array.jl:919 [inlined]
    [3] getindex
```

**Comment by jishnub** [comment]
it seems https://github.com/JuliaLang/julia/pull/53573 is needed as well

**Pull request merged**